### PR TITLE
test: Added goleak to TestMain in cmd/collector/app/server package folder

### LINKED ISSUE #5006: Enable and enforce goroutine leak checks in tests
Enforce the use of https://github.com/uber-go/goleak across the repo:
 * [x] a linter similar to scripts/check-test-files.sh that verifies that all packages with tests have goleak in TestMain
   * first deploy the script in audit mode where it only prints offending packages
   * then make the script fail the CI once the next item is addressed
   * Done in #5010
 * [x] add goleak to all packages via TestMain. Some packages may pass the validation, some may fail and would need to be fixed. It's best to break this into several PRs, first to add goleak to packages that do not fail, and then potentially separate PRs for each package where goleak does fail with the corresponding fixes to the tests.
   * Run `make goleak` to see the list of packages missing the linter
   * In packages with multiple `*_test.go` files add the check to a new file `package_test.go`
     * Remember to add copyright header - run `make fmt` after committing the new files
   * Examples: 
     * #5015 
     * #5017
 * [x] Fix leaks in the integration tests (see #6339)
 * [x] Change linter script to fail CI on errors going forward

State on March 8 2024:
```
$ make goleak
Verifying that all packages with tests have goleak in their TestMain
🔴 Error(check-goleak): no goleak check in package ./cmd/collector/app/
🔴 Error(check-goleak): no goleak check in package ./cmd/internal/printconfig/
🔴 Error(check-goleak): no goleak check in package ./cmd/query/app/
🔴 Error(check-goleak): no goleak check in package ./pkg/cassandra/gocql/testutils/
🔴 Error(check-goleak): no goleak check in package ./pkg/testutils/
🔴 Error(check-goleak): no goleak check in package ./plugin/metrics/prometheus/metricsstore/
🔴 Error(check-goleak): no goleak check in package ./plugin/sampling/strategystore/adaptive/
🔴 Error(check-goleak): no goleak check in package ./plugin/storage/integration/
Error(check-goleak): no goleak check in 8 package(s).
```

### PR DESCRIPTION
**Pull request by limistah**
## Which problem is this PR solving?
- Part of #5006 

## Description of the changes
- Added TestMain to package folders
- Added make goleak into unit test ci

## How was this change tested?
- please run make goleak 

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: .github/workflows/ci-unit-tests.yml [hunk 1]
```diff
@@ -35,6 +35,9 @@ jobs:
     - name: Run unit tests
       run: make test-ci
 
+    - name: Run go leak
+      run: make goleak
```
Comment (yurishkuro) [inline .github/workflows/ci-unit-tests.yml:39]: this is already called as part of the lint workflow, please remove

File: cmd/agent/app/reporter/grpc/grpc_test.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"testing"
+
+	"go.uber.org/goleak"
+)
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
+}
```
Comment (yurishkuro) [inline cmd/agent/app/reporter/grpc/grpc_test.go:25]:
```
$ go test ./cmd/agent/app/reporter/grpc/
PASS
goleak: Errors on successful test run: found unexpected goroutines:
```

File: cmd/collector/app/app_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"testing"
+
+	"go.uber.org/goleak"
+)
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
+}
```
Comment (yurishkuro) [inline cmd/collector/app/app_test.go:25]:
```
$ go test ./cmd/collector/app/
```
finds the leaks

**Review by yurishkuro — CHANGES_REQUESTED** [review]
I spot-checked a couple packages with added goleak checks and the tests now fail with actual goleaks. The objective of the ticket was not just add the checks and break the CI, but to fix the leaks and only then add the checks,

**Comment by yurishkuro** [comment]
btw the test workflow was successful here because of a recently introduced bug -- #5057

**Comment by limistah** [comment]
> I spot-checked a couple packages with added goleak checks and the tests now fail with actual goleaks. The objective of the ticket was not just add the checks and break the CI, but to fix the leaks and only then add the checks.

Cool, I'd check in the leaks.

Did a quick run on them and noticed a few of them, some were floated from vendor packages.

I might have to reduce the scope of the fix as the errors returned by the go test path/to/package returned quite a lot of them.

**Comment by yurishkuro** [comment]
> I might have to reduce the scope of the fix as the errors returned by the go test path/to/package returned quite a lot of them.

this was exactly the recommendation in the ticket.

**Comment by limistah** [comment]
> > I might have to reduce the scope of the fix as the errors returned by the go test path/to/package returned quite a lot of them.
> 
> this was exactly the recommendation in the ticket.

Great, I added the goleak fixes for `cmd/collector/app/server` package.

File: cmd/collector/app/server/grpc_test.go [hunk 4]
```diff
@@ -35,6 +36,13 @@ import (
 	"github.com/jaegertracing/jaeger/proto-gen/api_v2"
 )
 
+func stopOpenCensus(t *testing.T) {
```
Comment (yurishkuro) [inline cmd/collector/app/server/grpc_test.go:39]: this is wrong. We don't call opencensus so should not be stopping it manually either. It's most likely some other component starts it and we're not shutting down that component.

File: cmd/collector/app/server/http_test.go [hunk 5]
```diff
@@ -240,18 +248,21 @@ func TestSpanCollectorHTTPS(t *testing.T) {
 			} else {
 				require.NoError(t, requestError)
 				require.NotNil(t, response)
+				response.Body.Close()
```
Comment (yurishkuro) [inline cmd/collector/app/server/http_test.go:251]: this should be on L245 as `defer ...`

File: cmd/collector/app/server/http_test.go [hunk 6]
```diff
@@ -64,26 +64,30 @@ func TestCreateTLSHTTPServerError(t *testing.T) {
...
 	server := httptest.NewServer(nil)
-	defer server.Close()
 
 	serveHTTP(server.Config, server.Listener, params)
 
 	response, err := http.Post(server.URL, "", nil)
 	require.NoError(t, err)
 	assert.NotNil(t, response)
+	response.Body.Close()
+	server.Close()
```
Comment (yurishkuro) [inline cmd/collector/app/server/http_test.go:90]: why did you move it? `defer close` after creating is the correct pattern

File: cmd/collector/app/server/http_test.go (hunk 5 above)
Comment (limistah) [reply cmd/collector/app/server/http_test.go:251]: In a case when there is an error in the request, it throws a nil pointer since the request.Body might not have been initialized 

github.com/jaegertracing/jaeger/cmd/collector/app/server.serveHTTP.func1
        /Users/limistah/Work/oss/jaeger/cmd/collector/app/server/http.go:111
--- FAIL: TestSpanCollectorHTTPS (0.02s)
    --- FAIL: TestSpanCollectorHTTPS/should_fail_with_TLS_client_to_untrusted_TLS_server (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x40 pc=0x104e6664c]

File: cmd/collector/app/server/grpc_test.go (hunk 4 above)
Comment (limistah) [reply cmd/collector/app/server/grpc_test.go]: So, I removed the code.

The reason why I added it is that [opencensus has an init](https://github.com/census-instrumentation/opencensus-go/blob/01e6da5fc01c42aca1e0ce315f41744876e9fcbb/stats/view/worker.go#L33) that starts the worker 


I have searched for [NewMeter](https://github.com/census-instrumentation/opencensus-go/blob/01e6da5fc01c42aca1e0ce315f41744876e9fcbb/stats/view/worker.go#L261) within this codebase, the best that I could find is inside of the cmd/collector/app/handler using [opentelemtry's NewMeterProvider](https://pkg.go.dev/go.opentelemetry.io/otel/metric@v1.21.0/noop#NewMeterProvider) but none of these exposes the worker.

Here is a screenshot of the error:

image

Any suggestion here, I sincerely don't want to silent the leak.

File: cmd/collector/app/server/server_test.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2023 The Jaeger Authors.
```
Comment (yurishkuro) [inline cmd/collector/app/server/server_test.go:1]: we call these `package_test.go`

File: cmd/collector/app/server/grpc_test.go (hunk 4 above)
Comment (yurishkuro) [reply cmd/collector/app/server/grpc_test.go]: I prefer to silence it. Starting goroutines in init() is a stupid pattern and it does not indicate a leak in our code, so why bother handling it manually. See #5066 for similar issue.

I believe newer versions of OTEL Collector will soon remove opencensus dependency altogether.

Comment (limistah) [reply cmd/collector/app/server/grpc_test.go]: Fair enough

**Review by yurishkuro — APPROVED** [review]

File: pkg/testutils/leakcheck_test.go [hunk 8]
```diff
@@ -24,3 +24,10 @@ func TestIgnoreGlogFlushDaemonLeak(t *testing.T) {
 		t.Errorf("IgnoreGlogFlushDaemonLeak() returned nil, want non-nil goleak.Option")
 	}
 }
+
+func TestIgnoreOpenCensusWorkerLeak(t *testing.T) {
+	opt := IgnoreOpenCensusWorkerLeak()
+	if opt == nil {
+		t.Errorf("IgnoreOpenCensusWorkerLeak() returned nil, want non-nil goleak.Option")
+	}
```
Comment (yurishkuro) [inline pkg/testutils/leakcheck_test.go:32]:
```suggestion
	assert.NotNil(t, IgnoreOpenCensusWorkerLeak())
```

File: pkg/testutils/leakcheck_test.go [hunk 9]
```diff
@@ -24,3 +24,7 @@ func TestIgnoreGlogFlushDaemonLeak(t *testing.T) {
 		t.Errorf("IgnoreGlogFlushDaemonLeak() returned nil, want non-nil goleak.Option")
 	}
 }
+
+func TestIgnoreOpenCensusWorkerLeak(t *testing.T) {
+	assert.NotNil(t, IgnoreOpenCensusWorkerLeak())
```
Comment (yurishkuro) [inline pkg/testutils/leakcheck_test.go:29]: this will not compile, you need an import from stretchr/testify

**Review by yurishkuro — APPROVED** [review]

**Comment by limistah** [comment]
I ran make cover on [this commit](https://github.com/jaegertracing/jaeger/pull/5055/commits/2151038a6d9c95cb8ca4580f7918a1c529dd5aaf)

Which gave the below result for `pkg/testutils/leakcheck_test.go`
image

**Review by yurishkuro — APPROVED** [review]
